File: frontend/src/wasm/guard.ts
import {
  IDX_INBOX_DIRTY,
  IDX_OUTBOX_HOST_DIRTY,
  IDX_OUTBOX_KERNEL_DIRTY,
  IDX_MESH_EVENT_EPOCH,
  IDX_ARENA_ALLOCATOR,
  OFFSET_REGION_GUARDS,
  SIZE_REGION_GUARDS,
  REGION_GUARD_ENTRY_SIZE,
  REGION_GUARD_COUNT,
} from './layout';

export enum RegionOwner {
  Kernel = 1 << 0,
  Module = 1 << 1,
  Host = 1 << 2,
  System = 1 << 3,
}

export enum RegionId {
  Inbox = 0,
  OutboxHost = 1,
  OutboxKernel = 2,
  MeshEventQueue = 3,
  ArenaRequestQueue = 4,
  ArenaResponseQueue = 5,
  Arena = 6,
}

export enum AccessMode {
  ReadOnly = 0,
  SingleWriter = 1,
  MultiWriter = 2,
}

export type RegionPolicy = {
  regionId: RegionId;
  access: AccessMode;
  writerMask: number;
  readerMask: number;
  epochIndex?: number;
};

const GUARD_LOCK = 0;
const GUARD_LAST_EPOCH = 1;
const GUARD_VIOLATIONS = 2;
const GUARD_LAST_OWNER = 3;

export function policyFor(region: RegionId): RegionPolicy {
  switch (region) {
    case RegionId.Inbox:
      return {
        regionId: region,
        access: AccessMode.SingleWriter,
        writerMask: RegionOwner.Kernel,
        readerMask: RegionOwner.Module | RegionOwner.Host,
        epochIndex: IDX_INBOX_DIRTY,
      };
    case RegionId.OutboxHost:
      return {
        regionId: region,
        access: AccessMode.SingleWriter,
        writerMask: RegionOwner.Kernel,
        readerMask: RegionOwner.Host,
        epochIndex: IDX_OUTBOX_HOST_DIRTY,
      };
    case RegionId.OutboxKernel:
      return {
        regionId: region,
        access: AccessMode.MultiWriter,
        writerMask: RegionOwner.Module,
        readerMask: RegionOwner.Kernel,
        epochIndex: IDX_OUTBOX_KERNEL_DIRTY,
      };
    case RegionId.MeshEventQueue:
      return {
        regionId: region,
        access: AccessMode.SingleWriter,
        writerMask: RegionOwner.Kernel,
        readerMask: RegionOwner.Host,
        epochIndex: IDX_MESH_EVENT_EPOCH,
      };
    case RegionId.ArenaRequestQueue:
      return {
        regionId: region,
        access: AccessMode.SingleWriter,
        writerMask: RegionOwner.Module,
        readerMask: RegionOwner.Kernel,
        epochIndex: IDX_ARENA_ALLOCATOR,
      };
    case RegionId.ArenaResponseQueue:
      return {
        regionId: region,
        access: AccessMode.SingleWriter,
        writerMask: RegionOwner.Kernel,
        readerMask: RegionOwner.Module,
      };
    case RegionId.Arena:
      return {
        regionId: region,
        access: AccessMode.MultiWriter,
        writerMask: RegionOwner.Kernel | RegionOwner.Module,
        readerMask: RegionOwner.Kernel | RegionOwner.Module | RegionOwner.Host,
      };
    default:
      return {
        regionId: region,
        access: AccessMode.ReadOnly,
        writerMask: 0,
        readerMask: 0,
      };
  }
}

function guardIndex(regionId: number, field: number): number {
  const entryWords = REGION_GUARD_ENTRY_SIZE / 4;
  return regionId * entryWords + field;
}

function getGuardView(sab: SharedArrayBuffer): Int32Array {
  return new Int32Array(sab, OFFSET_REGION_GUARDS, SIZE_REGION_GUARDS / 4);
}

export function resetGuardTable(sab: SharedArrayBuffer): void {
  const view = getGuardView(sab);
  view.fill(0);
}

function incrementViolation(view: Int32Array, regionId: number): void {
  Atomics.add(view, guardIndex(regionId, GUARD_VIOLATIONS), 1);
}

export function validateRegionRead(
  sab: SharedArrayBuffer,
  region: RegionId,
  owner: RegionOwner
): boolean {
  const policy = policyFor(region);
  if ((policy.readerMask & owner) === 0) {
    const view = getGuardView(sab);
    incrementViolation(view, region);
    return false;
  }
  return true;
}

export class RegionGuard {
  private guardView: Int32Array;
  private flagsView: Int32Array | null;
  private policy: RegionPolicy;
  private owner: RegionOwner;
  private startEpoch?: number;
  private locked: boolean;

  constructor(
    guardView: Int32Array,
    flagsView: Int32Array | null,
    policy: RegionPolicy,
    owner: RegionOwner,
    locked: boolean
  ) {
    this.guardView = guardView;
    this.flagsView = flagsView;
    this.policy = policy;
    this.owner = owner;
    this.locked = locked;
    if (policy.epochIndex !== undefined && this.flagsView) {
      this.startEpoch = Atomics.load(this.flagsView, policy.epochIndex);
    }
  }

  ensureEpochAdvanced(): boolean {
    if (
      this.policy.epochIndex === undefined ||
      this.startEpoch === undefined ||
      !this.flagsView
    ) {
      return true;
    }
    const current = Atomics.load(this.flagsView, this.policy.epochIndex);
    if (current <= this.startEpoch) {
      incrementViolation(this.guardView, this.policy.regionId);
      return false;
    }
    Atomics.store(this.guardView, guardIndex(this.policy.regionId, GUARD_LAST_EPOCH), current);
    return true;
  }

  release(): boolean {
    if (!this.locked) {
      return true;
    }
    const idx = guardIndex(this.policy.regionId, GUARD_LOCK);
    const released =
      Atomics.compareExchange(this.guardView, idx, this.owner, 0) === this.owner;
    if (!released) {
      incrementViolation(this.guardView, this.policy.regionId);
    }
    this.locked = false;
    return released;
  }
}

export function acquireRegionWrite(
  sab: SharedArrayBuffer,
  flagsView: Int32Array | null,
  region: RegionId,
  owner: RegionOwner
): RegionGuard | null {
  const policy = policyFor(region);
  if ((policy.writerMask & owner) === 0) {
    const view = getGuardView(sab);
    incrementViolation(view, region);
    return null;
  }

  const view = getGuardView(sab);
  let locked = false;
  if (policy.access === AccessMode.SingleWriter) {
    const idx = guardIndex(region, GUARD_LOCK);
    locked = Atomics.compareExchange(view, idx, 0, owner) === 0;
    if (!locked) {
      incrementViolation(view, region);
      return null;
    }
  } else if (policy.access === AccessMode.MultiWriter) {
    Atomics.store(view, guardIndex(region, GUARD_LAST_OWNER), owner);
  }

  return new RegionGuard(view, flagsView, policy, owner, locked);
}

File: modules/sdk/src/guard.rs
use crate::layout::{
    OFFSET_REGION_GUARDS, REGION_GUARD_COUNT, REGION_GUARD_ENTRY_SIZE, SIZE_REGION_GUARDS,
};
use crate::sab::SafeSAB;

/// Region Owner bitmask (shared across Go/Rust/JS)
#[repr(u32)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RegionOwner {
    Kernel = 1 << 0,
    Module = 1 << 1,
    Host = 1 << 2,
    System = 1 << 3,
}

/// Access mode for a region
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AccessMode {
    ReadOnly,
    SingleWriter,
    MultiWriter,
}

/// Region policy: who can read/write and how access is enforced
#[derive(Clone, Copy, Debug)]
pub struct RegionPolicy {
    pub region_id: u32,
    pub access: AccessMode,
    pub writer_mask: u32,
    pub reader_mask: u32,
    pub epoch_index: Option<u32>,
}

/// Known regions for guard enforcement
#[repr(u32)]
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RegionId {
    Inbox = 0,
    OutboxHost = 1,
    OutboxKernel = 2,
    MeshEventQueue = 3,
    ArenaRequestQueue = 4,
    ArenaResponseQueue = 5,
    Arena = 6,
}

pub fn policy_for(region: RegionId) -> RegionPolicy {
    match region {
        RegionId::Inbox => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::SingleWriter,
            writer_mask: RegionOwner::Kernel as u32,
            reader_mask: RegionOwner::Module as u32 | RegionOwner::Host as u32,
            epoch_index: Some(crate::layout::IDX_INBOX_DIRTY),
        },
        RegionId::OutboxHost => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::SingleWriter,
            writer_mask: RegionOwner::Kernel as u32,
            reader_mask: RegionOwner::Host as u32,
            epoch_index: Some(crate::layout::IDX_OUTBOX_HOST_DIRTY),
        },
        RegionId::OutboxKernel => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::MultiWriter,
            writer_mask: RegionOwner::Module as u32,
            reader_mask: RegionOwner::Kernel as u32,
            epoch_index: Some(crate::layout::IDX_OUTBOX_KERNEL_DIRTY),
        },
        RegionId::MeshEventQueue => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::SingleWriter,
            writer_mask: RegionOwner::Kernel as u32,
            reader_mask: RegionOwner::Host as u32,
            epoch_index: Some(crate::layout::IDX_MESH_EVENT_EPOCH),
        },
        RegionId::ArenaRequestQueue => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::SingleWriter,
            writer_mask: RegionOwner::Module as u32,
            reader_mask: RegionOwner::Kernel as u32,
            epoch_index: Some(crate::layout::IDX_ARENA_ALLOCATOR),
        },
        RegionId::ArenaResponseQueue => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::SingleWriter,
            writer_mask: RegionOwner::Kernel as u32,
            reader_mask: RegionOwner::Module as u32,
            epoch_index: None,
        },
        RegionId::Arena => RegionPolicy {
            region_id: region as u32,
            access: AccessMode::MultiWriter,
            writer_mask: RegionOwner::Kernel as u32 | RegionOwner::Module as u32,
            reader_mask: RegionOwner::Kernel as u32
                | RegionOwner::Module as u32
                | RegionOwner::Host as u32,
            epoch_index: None,
        },
    }
}

#[derive(Debug)]
pub enum GuardError {
    Unauthorized(&'static str),
    Locked(&'static str),
    OutOfRange(&'static str),
}

/// Guard entry layout (4 x i32)
const GUARD_LOCK: u32 = 0; // Active writer lock (owner id)
const GUARD_LAST_EPOCH: u32 = 1;
const GUARD_VIOLATIONS: u32 = 2;
const GUARD_LAST_OWNER: u32 = 3;

fn guard_index(region_id: u32, field: u32) -> u32 {
    let entry_words = (REGION_GUARD_ENTRY_SIZE / 4) as u32;
    region_id * entry_words + field
}

fn increment_violation(sab: &SafeSAB, region_id: u32) {
    if let Ok(flags) = sab.int32_view(OFFSET_REGION_GUARDS, SIZE_REGION_GUARDS / 4) {
        let idx = guard_index(region_id, GUARD_VIOLATIONS);
        let _ = crate::js_interop::atomic_add(&flags, idx, 1);
    }
}

fn read_guard_word(sab: &SafeSAB, region_id: u32, field: u32) -> Option<u32> {
    sab.int32_view(OFFSET_REGION_GUARDS, SIZE_REGION_GUARDS / 4)
        .ok()
        .map(|flags| crate::js_interop::atomic_load(&flags, guard_index(region_id, field)) as u32)
}

fn write_guard_word(sab: &SafeSAB, region_id: u32, field: u32, value: u32) {
    if let Ok(flags) = sab.int32_view(OFFSET_REGION_GUARDS, SIZE_REGION_GUARDS / 4) {
        crate::js_interop::atomic_store(&flags, guard_index(region_id, field), value as i32);
    }
}

fn cas_guard_word(sab: &SafeSAB, region_id: u32, field: u32, expected: u32, value: u32) -> bool {
    if let Ok(flags) = sab.int32_view(OFFSET_REGION_GUARDS, SIZE_REGION_GUARDS / 4) {
        let actual =
            crate::js_interop::atomic_compare_exchange(&flags, guard_index(region_id, field), expected as i32, value as i32);
        return actual as u32 == expected;
    }
    false
}

pub struct RegionGuard {
    sab: SafeSAB,
    policy: RegionPolicy,
    owner: RegionOwner,
    start_epoch: Option<u32>,
    released: bool,
}

impl RegionGuard {
    pub fn acquire_write(sab: SafeSAB, policy: RegionPolicy, owner: RegionOwner) -> Result<Self, GuardError> {
        if (owner as u32) & policy.writer_mask == 0 {
            increment_violation(&sab, policy.region_id);
            return Err(GuardError::Unauthorized("writer not allowed for region"));
        }

        if policy.region_id >= REGION_GUARD_COUNT as u32 {
            return Err(GuardError::OutOfRange("region id out of range"));
        }

        match policy.access {
            AccessMode::ReadOnly => {
                increment_violation(&sab, policy.region_id);
                return Err(GuardError::Unauthorized("region is read-only"));
            }
            AccessMode::SingleWriter => {
                if !cas_guard_word(&sab, policy.region_id, GUARD_LOCK, 0, owner as u32) {
                    increment_violation(&sab, policy.region_id);
                    return Err(GuardError::Locked("region already locked"));
                }
            }
            AccessMode::MultiWriter => {
                // No lock, but record last owner for debugging
                write_guard_word(&sab, policy.region_id, GUARD_LAST_OWNER, owner as u32);
            }
        }

        let start_epoch = policy
            .epoch_index
            .map(|idx| crate::js_interop::atomic_load(sab.barrier_view(), idx) as u32);

        Ok(Self {
            sab,
            policy,
            owner,
            start_epoch,
            released: false,
        })
    }

    pub fn validate_read(sab: &SafeSAB, policy: RegionPolicy, owner: RegionOwner) -> Result<(), GuardError> {
        if (owner as u32) & policy.reader_mask == 0 {
            increment_violation(sab, policy.region_id);
            return Err(GuardError::Unauthorized("reader not allowed for region"));
        }
        Ok(())
    }

    /// Verify epoch advanced after a write (if configured).
    pub fn ensure_epoch_advanced(&self) -> Result<(), GuardError> {
        let Some(idx) = self.policy.epoch_index else {
            return Ok(());
        };
        let Some(start) = self.start_epoch else {
            return Ok(());
        };
        let current = crate::js_interop::atomic_load(self.sab.barrier_view(), idx) as u32;
        if current <= start {
            increment_violation(&self.sab, self.policy.region_id);
            return Err(GuardError::Unauthorized("epoch not advanced after write"));
        }
        write_guard_word(&self.sab, self.policy.region_id, GUARD_LAST_EPOCH, current);
        Ok(())
    }

    pub fn release(mut self) -> Result<(), GuardError> {
        if self.released {
            return Ok(());
        }

        if self.policy.access == AccessMode::SingleWriter {
            if !cas_guard_word(&self.sab, self.policy.region_id, GUARD_LOCK, self.owner as u32, 0) {
                increment_violation(&self.sab, self.policy.region_id);
                return Err(GuardError::Locked("release failed: lock owner mismatch"));
            }
        }

        self.released = true;
        Ok(())
    }
}

impl Drop for RegionGuard {
    fn drop(&mut self) {
        if self.released {
            return;
        }
        if self.policy.access == AccessMode::SingleWriter {
            let _ = cas_guard_word(&self.sab, self.policy.region_id, GUARD_LOCK, self.owner as u32, 0);
        }
    }
}
